File: ruby/trinary/trinary.rb
class Trinary
  DEFAULT_OUTPUT = 0

  attr_reader :digits

  def initialize(digits)
    @digits = digits
  end

  def to_decimal
    return DEFAULT_OUTPUT if invalid_digits?

    formatted_digits.each_with_index.inject(0) do |sum, (digit, index)|
      sum + convert_to_trinary(digit, index)
    end
  end
end

private

def convert_to_trinary(digit, index)
  digit.to_i * (3**index)
end

def formatted_digits
  digits.reverse.split('')
end

def invalid_digits?
  digits.match(/[^0-2]/)
end

module BookKeeping
  VERSION = 1
end

File: javascript/trinary/trinary.js
class Trinary {
  constructor(trinary) {
    this.trinary = trinary;
  }

  toDecimal() {
    if (/[^\d]+/.test(this.trinary)) return 0;

    return this.trinary
    .split('')
    .reverse()
    .map((digit, index) => parseInt(digit) * Math.pow(3, index))
    .reduce((a, b) => a + b, 0);
  }
}

module.exports = Trinary;
